Normalise bound-free estimators on each rank independently to eliminate MPI_Bcast and skip NLTE solver if Te=MINTEMP

jpollin98 opened the pull request:
**Normalise bound-free estimators independently on all ranks**
The normalisation of the bf estimators has been changed to occur across all cells instead of the rank-assigned cells. This means that we no longer need to broadcast the prev_bfratenormed array.

**Skip NLTE solver if Te=MINTEMP:**
The temperature check was added due to populations in some grid cells with a low temperature, as populations being unphysical (i.e producing nne outsude of the limits of a float). These grid cells typically had a lower amount of 56Ni.

jpollin98 commented:
Changes have been made to sections of the code which deal with the bfestimators. 

MPI_broadcast of the prev_bfrate_normed quantity was causing errors. As such, I have changed the do_MPI_Bcast to not broadcast the prev_bfrate_normed quantitity. This has meant that I have changed the radfield::normalise_bf_estimators to occur earlier across each rank, I this means that we dont have to broadcast the prev_bfrate_normed quantity anymore.

jpollin98 commented:
Regarding the movement of the normalise_bf_estimators, I have ran the W7 model and it produces consistent results. 

However, a mpirun error has appeared within the nt_MPI_Bcast() function when broadcasting the frac_deposition ratecoeffperdeposition and lineindex across all ranks. I think this error is happening because the broadcast is MPI_COMM_WORLD. Any advice on how to fix/change this part of the code would be appreciated.

lukeshingles commented:
> Changes have been made to sections of the code which deal with the bfestimators.
> 
> MPI_broadcast of the prev_bfrate_normed quantity was causing errors. As such, I have changed the do_MPI_Bcast to not broadcast the prev_bfrate_normed quantitity. This has meant that I have changed the radfield::normalise_bf_estimators to occur earlier across each rank, I this means that we dont have to broadcast the prev_bfrate_normed quantity anymore.

Can you put your description into the first comment, please? Only the first message will be kept after merging.

lukeshingles commented:
> Regarding the movement of the normalise_bf_estimators, I have ran the W7 model and it produces consistent results.
> 
> However, a mpirun error has appeared within the nt_MPI_Bcast() function when broadcasting the frac_deposition ratecoeffperdeposition and lineindex across all ranks. I think this error is happening because the broadcast is MPI_COMM_WORLD. Any advice on how to fix/change this part of the code would be appreciated.

To me, it sounds like either the cluster is failing or there are some invalid memory accesses in the code. The CI tests for the parts of the code that you're editing can easily miss MPI problems because the nebularonezone test is run with two ranks on a single node, and the model has only a single radial shell.

lukeshingles commented on radfield.cc:
```diff
@@ -1138,11 +1138,25 @@ void reduce_estimators()
   if constexpr (DETAILED_BF_ESTIMATORS_ON) {
     for (ptrdiff_t nonemptymgi = 0; nonemptymgi < nonempty_npts_model; nonemptymgi++) {
       MPI_Allreduce(MPI_IN_PLACE, &bfrate_raw[nonemptymgi * globals::bfestimcount], globals::bfestimcount, MPI_DOUBLE,
-                    MPI_SUM, MPI_COMM_WORLD);
+                    MPI_SUM, globals::mpi_comm_node);
```
The result will be identical, but if you're not going to use the reduction results across the whole node (you overwrite it a few lines down for all ranks except for rank_in_node == 0) then you can avoid communicating the unused results by replacing MPI_Allreduce with MPI_Reduce:

`
MPI_Reduce(my_rank == 0 ? MPI_IN_PLACE : &bfrate_raw[nonemptymgi * globals::bfestimcount], &bfrate_raw[nonemptymgi * globals::bfestimcount], globals::bfestimcount, MPI_DOUBLE, MPI_SUM, 0, globals::mpi_comm_node);
`

jpollin98 replied:
I ran with this version of the MPI_Allreduce on a cluster and ran into an error. I am running it again with the original code. I will investigate this error once the nt_MPI_Bcast error has been solved.

jpollin98 commented on nltepop.cc:
```diff
@@ -126,8 +126,8 @@ static void filter_nlte_matrix(const int element, gsl_matrix *rate_matrix, gsl_v
 static auto get_total_rate(const int index_selected, const gsl_matrix *rate_matrix, const gsl_vector *popvec,
-                           const bool into_level, const bool only_levels_below,
-                           const bool only_levels_above) -> double {
+                           const bool into_level, const bool only_levels_below, const bool only_levels_above)
+    -> double {
```
This is the problem, as I was using Clang 17. I have spent the last couple of hours trying to get Clang 18 installed and working. But GitHub cannot find it. It must be a pathing problem.

jpollin98 commented on nltepop.cc:
```diff
@@ -808,6 +808,16 @@ void solve_nlte_pops_element(const int element, const int modelgridindex, const
     return;
   }
 
+  double cell_Te = grid::get_Te(modelgridindex);
+
+  if (cell_Te < 1001.) {
+    printout(
+        "Not solving for NLTE populations in cell %d at timestep %d for element Z=%d due to low temperature Te=%g\n",
+        modelgridindex, timestep, atomic_number, cell_Te);
+    set_element_pops_lte(modelgridindex, element);
+    return;
+  }
+
```
This has been added when the code solves for populations in some grid cells with a low temperature, as populations were being produced that weren't physical (i.e producing nne outsude of the limits of a float). These grid cells typically had a lower amount of 56Ni. The rationale Stuart and I used is that if the temperatures are already at minimum, it does not make sense to solve for their populations as we aren't treating them correctly anyway.

jpollin98 replied:
I have now Explicitly added a check against MINTEMP

jpollin98 commented on nltepop.cc:
```diff
@@ -921,7 +931,8 @@ void solve_nlte_pops_element(const int element, const int modelgridindex, const
   // calculate the normalisation factors and apply them to the matrix
   // columns and balance vector elements
   gsl_vector *pop_norm_factor_vec = gsl_vector_calloc(nlte_dimension);
-  nltepop_matrix_normalise(modelgridindex, element, rate_matrix, pop_norm_factor_vec);
+  gsl_vector_set_all(pop_norm_factor_vec, 1.0);
+  // nltepop_matrix_normalise(modelgridindex, element, rate_matrix, pop_norm_factor_vec);
```
This does change the outputs for cells with low densities. It might just be that both methods describe the cells well in this test model.

I think it would be a good idea to put some cells in these tests which have extreme densities/compositions as this would allow us to test the ability of functions to cope with grid cells in 3D models.

lukeshingles commented:
> > > Regarding the movement of the normalise_bf_estimators, I have ran the W7 model and it produces consistent results.
> > > However, a mpirun error has appeared within the nt_MPI_Bcast() function when broadcasting the frac_deposition ratecoeffperdeposition and lineindex across all ranks. I think this error is happening because the broadcast is MPI_COMM_WORLD. Any advice on how to fix/change this part of the code would be appreciated.
> > 
> > 
> > To me, it sounds like either the cluster is failing or there are some invalid memory accesses in the code. The CI tests for the parts of the code that you're editing can easily miss MPI problems because the nebularonezone test is run with two ranks on a single node, and the model has only a single radial shell.
> 
> > > Regarding the movement of the normalise_bf_estimators, I have ran the W7 model and it produces consistent results.
> > > However, a mpirun error has appeared within the nt_MPI_Bcast() function when broadcasting the frac_deposition ratecoeffperdeposition and lineindex across all ranks. I think this error is happening because the broadcast is MPI_COMM_WORLD. Any advice on how to fix/change this part of the code would be appreciated.
> > 
> > 
> > To me, it sounds like either the cluster is failing or there are some invalid memory accesses in the code. The CI tests for the parts of the code that you're editing can easily miss MPI problems because the nebularonezone test is run with two ranks on a single node, and the model has only a single radial shell.
> 
> I think you are correct about this. When I ran the checks for integer overflows, I never checked when the NLTE solver turned on as this would take prohibitively long on my laptop.
> 
> I would guess that there is no way to get the flag -fsanitize=integer to work on the cluster? Which would mean I just have to go through the code slowly and check for the locations of the error.

I have tried running a full-scale simulation in testmode (make TESTMODE=ON sn3d), but it slowed to a crawl on the atomic data and I gave up waiting for it. I suggest you try that first, and it it isn't making any progress, then just add -fsanitize=interger,address to the CXXFLAGS (e.g., where -std=c++20 is set).

jpollin98 commented:
> > > > Regarding the movement of the normalise_bf_estimators, I have ran the W7 model and it produces consistent results.
> > > > However, a mpirun error has appeared within the nt_MPI_Bcast() function when broadcasting the frac_deposition ratecoeffperdeposition and lineindex across all ranks. I think this error is happening because the broadcast is MPI_COMM_WORLD. Any advice on how to fix/change this part of the code would be appreciated.
> > > 
> > > 
> > > To me, it sounds like either the cluster is failing or there are some invalid memory accesses in the code. The CI tests for the parts of the code that you're editing can easily miss MPI problems because the nebularonezone test is run with two ranks on a single node, and the model has only a single radial shell.
> > 
> > 
> > > > Regarding the movement of the normalise_bf_estimators, I have ran the W7 model and it produces consistent results.
> > > > However, a mpirun error has appeared within the nt_MPI_Bcast() function when broadcasting the frac_deposition ratecoeffperdeposition and lineindex across all ranks. I think this error is happening because the broadcast is MPI_COMM_WORLD. Any advice on how to fix/change this part of the code would be appreciated.
> > > 
> > > 
> > > To me, it sounds like either the cluster is failing or there are some invalid memory accesses in the code. The CI tests for the parts of the code that you're editing can easily miss MPI problems because the nebularonezone test is run with two ranks on a single node, and the model has only a single radial shell.
> > 
> > 
> > I think you are correct about this. When I ran the checks for integer overflows, I never checked when the NLTE solver turned on as this would take prohibitively long on my laptop.
> > I would guess that there is no way to get the flag -fsanitize=integer to work on the cluster? Which would mean I just have to go through the code slowly and check for the locations of the error.
> 
> I have tried running a full-scale simulation in testmode (make TESTMODE=ON sn3d), but it slowed to a crawl on the atomic data and I gave up waiting for it. I suggest you try that first, and it it isn't making any progress, then just add -fsanitize=interger,address to the CXXFLAGS (e.g., where -std=c++20 is set).

Hi Luke, I tried running with the other modules you suggested and still had problems. When I added the MPI_Barrier(MPI_COMM_WORLD) statements, I got slightly further by about ~150 grid cells (~150000 more broadcasts). I have had a look at the values still trying to be broadcast. They are still sensible numbers, so I still feel like this is an MPI problem. 

I have changed the nt_MPI_Bcast broadcast to use MPI_Pack and MPI_Unpack instead. These have passed the GitHub tests, but would you be able to have a quick look at the code before I run the simulation? (Just in case there are any glaring issues.)

[Also I will tidy it up once I know the fix works]

jpollin98 commented:
> > > > Hi Luke, I tried running with the other modules you suggested and still had problems. When I added the MPI_Barrier(MPI_COMM_WORLD) statements, I got slightly further by about ~150 grid cells (~150000 more broadcasts). I have had a look at the values still trying to be broadcast. They are still sensible numbers, so I still feel like this is an MPI problem.
> > > > I have changed the nt_MPI_Bcast broadcast to use MPI_Pack and MPI_Unpack instead. These have passed the GitHub tests, but would you be able to have a quick look at the code before I run the simulation? (Just in case there are any glaring issues.)
> > > 
> > > 
> > > Try a rebase onto the current master. Fionn's deflagration models slowed to a crawl with some kind of bound-free process issue. The latest version is slower generally, but seems to avoid this kind of deadlock. See if it makes a difference for your models.
> > 
> > 
> > So, whatever issue affected Classic also affected the NLTE branch of the code? That is very surprising... I wouldn't have expected any specific issues with Classic to cause a problem in the MPI routines.
> 
> All of my NLTE models worked fine. I'm not sure exactly sure what caused the photoionisation issues in the classic model, but I can't say for sure that they didn't affect any NLTE models. Better to rebase onto the latest commit to save yourself time.

I have rebased my branch onto the new main version of the code. From the looks of it, my change, the normalisation of the pop_norm_factor_vec, now breaks the GitHub actions. I have reverted to the previous normalisation method (i.e. normalised to LTE). The temperature check may catch some of the most extreme populations, causing some of the errors.

However, I do not think how we normalise should change the solution as, in both cases, they should both be valid. I will run two nebular versions of the W7 Model, one with the normalisation and one without, to test if the change is significant. 

Also, once these normalisation issues have been checked, can we merge these into the main branch so I can open a separate PR for the NT_Bcast issues if needed?
